File: content/problems/1416-restore-the-array/solutions/suffix-partition-dp/solution.cpp
class Solution {
    public: int numberOfArrays(string text,int k){
        int n=text.size(),mod=1000000007;
        vector<int> dp(n+1);
        dp[n]=1;
        for(int i=n-1;i>=0;--i){
            if(text[i]=='0')continue;
            long long value=0;
            for(int j=i;j<n;++j){
                value=value*10+text[j]-'0';
                if(value>k)break;
                dp[i]=(dp[i]+dp[j+1])%mod;
            }
        }
        return dp[0];
    }
};

File: content/problems/1416-restore-the-array/solutions/suffix-partition-dp/Solution.kt
class Solution {
    fun numberOfArrays(text:String,k:Int):Int{
        val n=text.length;
        val mod=1_000_000_007;
        val dp=IntArray(n+1);
        dp[n]=1;
        for(i in n-1 downTo 0){
            if(text[i]=='0')continue;
            var value=0L;
            for(j in i until n){
                value=value*10+(text[j]-'0');
                if(value>k)break;
                dp[i]=(dp[i]+dp[j+1])%mod
            }
        };
        return dp[0]
    }
}
